log4j-api/src/main/java/org/apache/logging/log4j/spi/internal/DefaultScopedContextProvider.java

```diff
@@ -0,0 +1,399 @@
+            } else {
+                return new Instance(this, key, value);
+            }
+        } else {
+            if (getContext().isPresent()) {
+                final Map<String, ?> map = getContextMap();
+                map.remove(key);
+                return new Instance(this, map);
+            }
+        }
+        return newScopedContext();
+    }
+
+    /**
+     * Creates a ScopedContext Instance with a Map of keys and values.
+     * @param map the Map.
+     * @return the ScopedContext Instance constructed.
+     */
+    @Override
+    public ScopedContext.Instance newScopedContext(final Map<String, ?> map) {
+        if (map != null && !map.isEmpty()) {
+            final Map<String, Object> objectMap = new HashMap<>();
+            if (getContext().isPresent()) {
+                objectMap.putAll(getContext().get().contextMap);
+            }
+            map.forEach((key, value) -> {
+                if (value == null || (value instanceof String && ((String) value).isEmpty())) {
+                    objectMap.remove(key);
+                } else {
+                    objectMap.put(key, value);
+                }
+            });
+            return new Instance(this, objectMap);
+        } else {
+            return getContext().isPresent() ? getContext().get() : null;
+        }
+    }
+
+    private static void setupContext(
+            final Map<String, Object> contextMap,
+            final Map<String, String> threadContextMap,
+            final Collection<String> contextStack,
+            final Instance context) {
+        Instance scopedContext = context;
+        // If the current context has a Map then we can just use it.
+        if (context.contextMap == null) {
+            do {
+                if (scopedContext.contextMap != null) {
+                    // Once we hit a scope with an already populated Map we won't need to go any further.
+                    contextMap.putAll(scopedContext.contextMap);
+                    break;
+                } else if (scopedContext.key != null) {
+                    contextMap.putIfAbsent(scopedContext.key, scopedContext.value);
+                }
+                scopedContext = scopedContext.parent;
+            } while (scopedContext != null);
+            scopedContext = new Instance(context.getProvider(), contextMap);
+        }
+        if (threadContextMap != null && !threadContextMap.isEmpty()) {
+            ThreadContext.putAll(threadContextMap);
+        }
+        if (contextStack != null) {
+            ThreadContext.setStack(contextStack);
+        }
+        context.getProvider().addScopedContext(scopedContext);
+    }
+
+    private static final class Instance implements ScopedContext.Instance {
+
+        private final DefaultScopedContextProvider provider;
+        private final Instance parent;
+        private final String key;
+        private final Object value;
+        private final Map<String, ?> contextMap;
```

Comment:
In the default version the parent is necessary since we aren't using a queue in the ThreadLocal. Having the key and value eliminates having to create and copy maps for every call to the where method. We really only populate the Map on a run or call. 
Are you proposing we always allocate a Map? Yes, that gets rid of null checks but then requires copying the map. I would opt for the null checks over that.

Reply:
Sorry. This makes absolutely no sense to me. In the first case you are creating 3 maps, 2 of which can't be accessed by anything. Why?  The second case is more verbose.

I like the way it works now. It is clean and simple.

Reply:
Point taken.

Would you consider having a `KeyValueInstance` class with `parent`, `key` and `value` and a `MapInstance` with `contextMap`? I think it would improve the readability of the code. From my perspective it took me some time to understand why you don't have null checks when you call `getContext().get().contextMap.get(key)`. By splitting the classes it would be obvious to anyone that `MapInstance#contextMap` is not null.

Reply:
This suggestion is included in the #2450 proposal.

Reply:
I have implemented a MapInstance and KeyValueInstance.